### DISCUSSION TITLE
Has the OW 1.9 debugger inability to break anywhere in Windows 10 x64 been fixed ?

### DISCUSSION THREAD

**Discussion by lynnmcguire** [open-watcom/open-watcom-v2#828] [Q&A]
It looks like http://www.openwatcom.org/ is dead.  I cannot figure out how to ask any questions there.

Has the OW 1.9 debugger inability to break anywhere in Windows 10 x64 been fixed in OW 2.0 ?

Thanks,
Lynn
LMC@winsim.com

**Comment by jmalak** [comment]
There is no big difference in debug server between OW1.9 and OW 2.0 that probably same problem exists in OW 2.0.
Please specify more exactly what you mean by **"break anywhere"**

**Reply by lynnmcguire** [reply]
In other words, I set a breakpoint on a function or at a line of code and the OW debugger goes right past the breakpoint, never stopping.   The OW 1.9 debugger works fine on Windows XP and Windows 7 x64 PCs.

**Reply by jmalak** [reply]
I am using some time OW 1.9 debugger on Windows 10 and break points works as expected.
May be it depends on debug information amount (what -d.. option you are using).

**Reply by lynnmcguire** [reply]
Are you using Windows 10 x86 or Windows 10 x64 ?  And I am using the Pro version of Windows 10 x64.

**Reply by lynnmcguire** [reply]
I am using /d2 for the Open Watcom C, C++, and F7 compilers.  Worked fine on Windows XP and 7.

**Reply by jmalak** [reply]
I am using Windows 10 x64. 
But I use OW1.9 debugger to debug OW2.0 generated code (32-bit).

**Reply by lynnmcguire** [reply]
So you can set a breakpoint and OW 1.9 debugger will stop at that breakpoint ?

**Reply by jmalak** [reply]
below is picture from debugging OW2.0 C compiler by OW1.9 debuger on Windows 10 x64.
I setup multiple breakpoints and press F5 or click on Go and it skip to next breakpoint.
<img width="950" alt="image" src="https://user-images.githubusercontent.com/2017260/155437070-64340587-2bfd-488d-a92d-2c45951d552d.png">

**Reply by lynnmcguire** [reply]
Ok, I am perplexed and confused.  I have wdw.exe in c:\watcom\binnt;

I do have wdw.exe set for Windows 7 compatibility mode.  I wonder if I should try a few more modes.

I am building huge programs, 850,000 lines of F77 code, 20,000 lines of C and C++ code.  All in a Win32 DLL.

**Reply by jmalak** [reply]
I don't have setup any compatibility mode for wdw on Windows 10 simply run it.

It could be a problem with size of debug info etc. Hard to say.
Also OW 2.0 has some OW 1.9 bug fixes.
Try to compile your project by OW 2.0 and try to debug it with OW 1.9 debugger and you will see if same problem or not.
OW 2.0 debugger has some problems with GUI part which is little reworked for 16/32/64-bit compatible code, it must be fixed.

**Reply by lynnmcguire** [reply]
I take it that there is no single file setup.exe for installing OW 2.0 ?

**Reply by jmalak** [reply]
> I take it that there is no single file setup.exe for installing OW 2.0 ?

I don't know what you are trying. You can download OW 2.0 installer, but if you don't want to change existing configuration then better is download OW 2.0 snapshot and use batch file to setup OW2.0 from command line.
contents of such batch file is standard as described in readme.
Below is sample for 64-bit Windows.

Change line 
SET WATCOM=C:\WATCOM
to point to your OW2.0 folder, where you download and extract OW2.0 distribution
```
---------------------------------------------------------------------------
@ECHO OFF
SET WATCOM=C:\WATCOM
SET PATH=%WATCOM%\BINNT64;%WATCOM%\BINNT;%PATH%
SET EDPATH=%WATCOM%\EDDAT
SET INCLUDE=%WATCOM%\H;%WATCOM%\H\NT
REM SET LIB=
REM SET WHTMLHELP=D:\BINNT\HELP
---------------------------------------------------------------------------
```

**Reply by lynnmcguire** [reply]
I have huge Win32 DLLs (10 MB and 30 MB).  I wonder if I am running past some internal limit for breakpoints on Windows 10 x64 WOW ?

**Reply by lynnmcguire** [reply]
BTW, if I a use a small win32 executable then the debugger break points work just fine.  So the problem is either that all of the breakpoints are in my DLLs or that my Win32 DLLs are so large (10 MB and 30 MB).

These breakpoints work just fine on my Windows 7 x64 PC.

**Reply by jmalak** [reply]
It is strange. Breakpoints are handled by hardware (4 breakpoint registers) and by `int 3` instruction.
May be a problem with authorities on Windows 10 that you need some kind of authorities to be able fully debug DLLs.
Try to do debug under administrator.

**Reply by lynnmcguire** [reply]
Yes, I have tried to debug as Administrator and the problem does not go away.

I have found a Fortran code problem and am fixing the 190 places that function is called.  Maybe when I am done the debugger will work properly.

**Reply by jmalak** [reply]
OK.
